### ISSUE TITLE
Workshop Notebooks

### ISSUE THREAD

**Issue by lesserwhirls** [Unidata/pyaos-ams-2020#13]
The notebooks currently in this repository were pulled from the notebook gallery (python-gallery repo) and act as placeholders. We need to decide what we will present in the 1 hour 45 minutes of time that we have. With those time constraints, and the fact that the range in skills of the students will be from completely new to python through expert, I think we consider this an interactive tour of the kinds of atmospheric science things you can do with python. Be sure to provide pleanty of pointers to https://unidata.github.io/python-training/, and feel free to invoke magic while highlighting the underlying libraries that make that magic happen.

If you plan on using a notebook that already exists, it would be a good idea to pull material from https://github.com/Unidata/python-training (the other workshop/gallery/etc repositories are deprecated in favor of training). @dopplershift gave me the heads up that the python-training repo notebooks contain fixes for the latest and greatest, and based on his twitter feed, I don't think we want to rediscover any lingering issues in, say, python-gallery :-)

I think it would be good to start with some basics, but unfortunately there are a lot of basics. I think the main idea I'd like those completely new to python to walk away with from an intro notebook would be that Numpy deals with arrays and matplotlib handles basic plotting, and cartopy does maps. I'd like to show enough such that they could plot, say, <code>y = x<sup>2</sup></code>, as well as points on a map using lat/lon pairs. This notebook will be something new, but at least we have so much material that it can be a create of Frankenstein, so to say. Those familiar with python might even get something out of the Cartopy example.

@jthielen has mentioned something featuring xarray. I think this would be great to feature, and builds naturally on "Numpy deals with arrays". We could add in the idea of units support here as well. He also floated the idea of the [Isentropic Analysis notebook](https://github.com/Unidata/python-training/blob/master/pages/workshop/MetPy_Advanced/Isentropic%20Analysis.ipynb) from the Advanced Metpy section of the workshops. That one might be too much material as it stands now, but I think getting to the end result would be worthwhile. This would also introduce the idea of reading remote data. Do you think you could come up with a notebook that does xarray, units, and siphon that revolves around isentropic analysis?

@mgrover1 has mentioned a Satellite declarative plotting example, possibly updated with more declarative plotting features. I think the declarative interface would be a good notebook to end with.

If we have three main notebooks, that'd be roughly 25 minutes each (providing 30 minutes for an intro and unexpected issues). I don't know if we can get another notebook in or not.

After the final notebook, we could go through what is available on https://unidata.github.io/python-training/ in a bit more detail (surfing the site only though), point out that many of us are here all week and (speaking for the Unidata people) we're at AMS to help, and highlight how to get support once AMS is finished.

Thoughts?

**Comment by jthielen** [comment]
@lesserwhirls, I think this is a great overall plan! It gives a nice overall structure to the workshop while striking a good balance that should benefit most of the audience.

> Do you think you could come up with a notebook that does xarray, units, and siphon that revolves around isentropic analysis?

Yes, I should be able to do so. I'm thinking a combination of the [xarray with MetPy tutorial from the MetPy docs](https://unidata.github.io/MetPy/latest/tutorials/xarray_tutorial.html) and the [isentropic analysis notebook](https://github.com/Unidata/python-training/blob/master/pages/workshop/MetPy_Advanced/Isentropic%20Analysis.ipynb), trimmed for 25 minutes of interactive teaching time, should work well?

> @mgrover1 has mentioned a Satellite declarative plotting example, possibly updated with more declarative plotting features. I think the declarative interface would be a good notebook to end with.

Are these the satellite examples from python-training? It looks like @jrleeman has been working on some recent improvements: https://github.com/Unidata/python-training/pull/35.

**Comment by lesserwhirls** [comment]
> Yes, I should be able to do so. I'm thinking a combination of the [xarray with MetPy tutorial from the MetPy docs](https://unidata.github.io/MetPy/latest/tutorials/xarray_tutorial.html) and the [isentropic analysis notebook](https://github.com/Unidata/python-training/blob/master/pages/workshop/MetPy_Advanced/Isentropic%20Analysis.ipynb), trimmed for 25 minutes of interactive teaching time, should work well?

Sounds like a plan to me!

> Are these the satellite examples from python-training? It looks like @jrleeman has been working on some recent improvements: [Unidata/python-training#35](https://github.com/Unidata/python-training/pull/35).

Not sure - @mgrover1?

**Comment by mgrover1** [comment]
@lesserwhirls Thanks for putting together this plan! Looks great! I will take a look at the PR that @jrleeman submitted. I just want to make sure to include all of the different types of declarative objects (`ImagePlot`, `PointObs`, `MapPanel`, and `PanelContainer`). This would also cover the new METAR support!

**Comment by mgrover1** [comment]
@lesserwhirls for the satellite, surface, and model data, should we create a data directory? That way the data is stored locally? The METAR LDM data feed is not always the most reliable, and it would be nice to have the satellite date/time lineup with the METAR/model data.

**Comment by dopplershift** [comment]
The only downside to a static data directory is that they're not learning skills that allow them to go home and get real data tomorrow. That may not be a goal for you all for this workshop. Just my $0.02.

**Comment by julienchastang** [comment]
A couple of notes:

- Environment

The environment accessible to users on pyaos-workshop.unidata.ucar.edu corresponds to the environment used for the Python gallery. You can double check this by examining the `environment.yml` in this repo.

If you check in notebooks and want to ensure the user is landed in the correct kernel from the get go, ensure the kernel spec region of the notebook looks like this:

```yaml
  "metadata": {
    "kernelspec": {
      "display_name": "Python [conda env:pyaos-ams-2020]",
      "language": "python",
      "name": "conda-env-pyaos-ams-2020-py"
    },
```

- Data

I would caution against storing (large) data files in this repo. The reason is that I would like to keep this repo with as small a footprint as possible to ensure users have a smoother experience upon login. Note that when the user logs in, this project will be cloned into their environment.

I may suggest using RAMADDA instead: http://ramadda-jetstream.unidata.ucar.edu/repository. With RAMADDA you can upload a CDM data file and get the DAP link for use in your notebook. Let me know if you want a login here and that can be arranged.

**Comment by mgrover1** [comment]
> The only downside to a static data directory is that they're not learning skills that allow them to go home and get real data tomorrow. That may not be a goal for you all for this workshop. Just my $0.02.

@dopplershift Do you have suggestions for reading in METAR data from the thredds-test server? As of now, I am using the `wget` module to read in the most recent surface observations.

Siphon can be used to read in the model data/satellite observations, but it is the surface data I am struggling with since often times, it requires downloading the file locally. It would not be a very large file (a couple megabytes).

**Comment by lesserwhirls** [comment]
I think trying to get real-time data is best. As a backup, we could include, but comment out, links to data sources that live in the casestudies:

https://thredds-test.unidata.ucar.edu/thredds/catalog/casestudies/catalog.html

Are there METAR reports in any of those?

**Comment by dopplershift** [comment]
@lesserwhirls There's a ton sitting in my [NOAAPORT Text Products collection](https://thredds-test.unidata.ucar.edu/thredds/catalog/noaaport/text/catalog.html). If you don't nuke the directory on the server, even if the realtime data feed goes down for some reason,  you should be good.

@mgrover1 We should be able to do the following (but sadly can't):
```python
from datetime import datetime
from metpy.io import parse_metar_file
from siphon.catalog import TDSCatalog

cat = TDSCatalog('https://thredds-test.unidata.ucar.edu/thredds/catalog/noaaport/text/metar/catalog.xml')
ds = cat.datasets.filter_time_nearest(datetime(2020, 1, 6, 14))
df = parse_metar_file(ds.remote_open())
```
That's the dream, but it's broken because I'm stupid and `remote_open()` assumes it should be doing binary I/O. It breaks on splitting on `'\n'`. :sob: So your next best option is to download the file then open:
```python
cat = TDSCatalog('https://thredds-test.unidata.ucar.edu/thredds/catalog/noaaport/text/metar/catalog.xml')
ds = cat.datasets.filter_time_nearest(datetime(2020, 1, 6, 14))
ds.download()
df = parse_metar_file(ds.name)
```
Now, the 0.12 release is not forgiving with Unicode failures, so you might hit some files that fail. This will be fixed in 1.0. If *absolutely need to* I can get a 0.12.1 with the fix (@kgoebber already made a PR)--but to me that would depend on how many files fail.

**Comment by lesserwhirls** [comment]
Quick update - @mwilson14 is going to present the opening notebook instead of me. I think it will be nice to have people who are still "in the trenches" with audience members, so to say, actually covering the notebooks.
